### PR TITLE
Clean up logic for guessing `-B` and `--lib`

### PR DESCRIPTION
**Pull request by harpocrates**
Haddock built with the `in-ghc-tree` flag tries harder to find the GHC lib folder and its own resources. This should make it possible to use `in-ghc-tree`-built Haddock without having to
specify the `-B` and `--lib` options.

### CODE REVIEW

**Comment by harpocrates** [comment]
Note to self: before merging remember to

  * update README to explain Haddock's development model (and at least mention `ghc-head`)
  * test rigorously that I haven't broken any other workflows with these changes

**Comment by sjakobi** [comment]
This looks like it could be helpful for me when working on Hi Haddock!

> test rigorously that I haven't broken any other workflows with these changes

I guess the most rigorous testing this will get will come after the merge! :grin:

**Comment by hvr** [comment]
@harpocrates btw, are you keeping in mind that the `in-ghc-tree` cabal flag is scheduled for removal?

File: haddock-api/src/Haddock.hs [hunk 1]
```diff
@@ -504,51 +499,79 @@ unsetPatternMatchWarnings dflags =
...
 #ifdef IN_GHC_TREE
-    ghcPath = "not available"
-#else
-    ghcPath = GhcPaths.ghc
-#endif
 
+-- | See 'getBaseDir' in "SysTools.BaseDir"
+getBaseDir :: IO (Maybe FilePath)
+getBaseDir = do
+  exec_path <- getExecutablePath
```
Comment (harpocrates) [inline haddock-api/src/Haddock.hs]: @hvr Is this safe across all platforms?

The previous version of this didn't use `getExecutablePath`. Additionally, I've seen other code where uses of `getExecutablePath` are limited via CPP to only some platforms (darwin, windows, and linux I think), but I can't find any justification for this.

Comment (harpocrates) [reply haddock-api/src/Haddock.hs]: Couple points from discord convo, and my resolution of them:

  * __Windows might be relying on this working, due to some issues preventing wrapper scripts.__
    I've double checked and `getExecutablePath` in `System.Environment.ExecutablePath` is just a slightly better take on the old `getExecDir` (same call to `GetModuleFileNameW`), so I doubt we'll regress here.
  * __`getExecutablePath` isn't reliable.__
    Looking again at the implementation, we should be ready to catch  exceptions throw by `getBaseDir` and return `Nothing` in those cases.

  * __Even if it returns something, we should validate it.__
    Checking for the existence of the `lib` directory seems like a best effort here. Good enough for now.

**Comment by harpocrates** [comment]
Cleaned everything up in a branch pushed to the official repo and opened a new PR: https://github.com/haskell/haddock/pull/1026.

**Pull request closed without merging**